File: solutions/python/robot-name/1/robot_name.py
import random
import string

class Robot:
    def __init__(self):
        self._set_name()
        self._set_name()

    def _set_name(self):
        letter1 = string.ascii_uppercase[random.randint(0, 25)]
        letter2 = string.ascii_uppercase[random.randint(0, 25)]
        self.name = letter1 + letter2 + str(random.randint(1, 9)) + str(random.randint(1, 9)) + str(random.randint(1, 9))

    def reset(self):
        self._set_name()

File: solutions/csharp/robot-name/1/RobotName.cs
public class Robot
{
  private static string _characters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
  private static List<string> _usedNames = new List<string>();
  private static Random _random = new Random();
  private string _name = string.Empty;

  public string Name
  {
    get
    {
      if (string.IsNullOrEmpty(_name))
      {
        _name = GenerateName();
      }
      return _name;
    }
  }

  private string GenerateName()
  {
    var name = string.Empty;
    do
    {
      name = _characters[_random.Next(0, 26)].ToString();
      name += _characters[_random.Next(0, 26)].ToString();
      name += _random.Next(0, 10).ToString("D");
      name += _random.Next(0, 10).ToString("D");
      name += _random.Next(0, 10).ToString("D");
    } while (_usedNames.Contains(name));
    _usedNames.Add(name);
    return name;
  }

  public void Reset()
  {
    _name = string.Empty;
  }
}